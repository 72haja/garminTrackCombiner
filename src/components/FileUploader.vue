<script setup>
import { ref, watch, computed } from 'vue';

const emit = defineEmits(['files-changed'])

const files = ref([])

function emitFilesChanged () {
  emit('files-changed', files.value || [])
}

function removeFile (fileName) {
  files.value = files.value?.filter((localFile) => localFile.name !== fileName) || []
}

const fileLength = computed(() => files.value?.length || 0)

watch(files, async () => {
  emitFilesChanged();
})
</script>

<template>
  <div>
    <v-file-input accept=".tcx"
                  label="TXC-Dateien"
                  outlined
                  chips
                  counter
                  :key="fileLength"
                  clearable
                  v-model="files"
                  class="w-100"
                  multiple>
      <template #selection="item">
        <v-chip v-for="(file, index) in item.fileNames"
                :key="index"
                closable
                @click:close.stop="removeFile(file)">
          {{ file }}
        </v-chip>
      </template>
    </v-file-input>
  </div>
</template>
<style lang='scss'>
</style>
