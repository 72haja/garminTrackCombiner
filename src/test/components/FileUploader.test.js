import { describe, expect, vi } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'

import { createVuetify } from 'vuetify'

import testFile1 from '../../../public/mocks/testFile1.tcx?raw'
import testFile2 from '../../../public/mocks/testFile2.tcx?raw'

import FileUploader from '../../components/FileUploader.vue'

describe('mount App', () => {
  const vuetify = createVuetify()

  test('mount component', async () => {
    expect(FileUploader).toBeTruthy()

    const wrapper = mount(FileUploader, {
      global: {
        plugins: [vuetify]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('test computed', () => {
  const vuetify = createVuetify()
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(FileUploader, {
      global: {
        plugins: [vuetify]
      }
    });
  });

  describe('fileLength', () => {
    test('fileLength - 0', async () => {
      wrapper.vm.files = null

      expect(wrapper.vm.fileLength).toBe(0)
    })

    test('errorMessageLength - 1 - exists', async () => {
      const testFile =
        new File(["Foo"], "testFile1", {
          type: "text/xml",
        })

      wrapper.vm.files = [
        testFile
      ]

      expect(wrapper.vm.fileLength).toBe(1)
    })
  });
})

describe('test function', () => {
  const vuetify = createVuetify()
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(FileUploader, {
      global: {
        plugins: [vuetify]
      }
    });
  });

  describe('emitFilesChanged', () => {
    test('emitFilesChanged', async () => {
      const testFile =
        new File(["Foo"], "testFile1", {
          type: "text/xml",
        })

      wrapper.vm.files = [
        testFile
      ]

      wrapper.vm.emitFilesChanged()

      await wrapper.vm.$nextTick();

      expect(wrapper.emitted('files-changed')).toBeTruthy();
      expect(wrapper.emitted('files-changed')[0]).toEqual([[
        testFile,
      ]]);
    })

    test('emitFilesChanged - no files', async () => {
      wrapper.vm.files = null

      wrapper.vm.emitFilesChanged()

      expect(wrapper.emitted('files-changed')).toBeTruthy();
      expect(wrapper.emitted('files-changed')[0]).toEqual([[]]);
    })
  })

  describe('removeFile', () => {
    it('removeFile', async () => {
      wrapper.vm.files = [
        new File(["Foo"], "testFile1", {
          type: "text/xml",
        }),
        new File([testFile2], "testFile2", {
          type: "text/xml",
        }),
      ]

      wrapper.vm.removeFile(wrapper.vm.files[0].name)

      expect(wrapper.vm.files).toEqual([
        new File([testFile2], "testFile2", {
          type: "text/xml",
        }),
      ])
    })

    it('removeFile - no files', async () => {
      wrapper.vm.files = null

      wrapper.vm.removeFile("not existing file name")

      expect(wrapper.vm.files).toEqual([])
    })
  });
})

