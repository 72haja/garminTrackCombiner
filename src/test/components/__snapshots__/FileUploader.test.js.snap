// Vitest Snapshot v1

exports[`mount App > mount component 1`] = `
"<div>
  <div class=\\"v-input v-input--horizontal v-input--density-default v-file-input w-100\\">
    <div class=\\"v-input__prepend\\"><i class=\\"mdi-paperclip mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable\\" aria-hidden=\\"true\\"></i>
      <!---->
    </div>
    <div class=\\"v-input__control\\">
      <div class=\\"v-field v-field--appended v-field--has-label v-field--variant-filled v-theme--light\\" prepend-icon=\\"$file\\">
        <div class=\\"v-field__overlay\\"></div>
        <div class=\\"v-field__loader\\">
          <div class=\\"v-progress-linear v-theme--light\\" style=\\"height: 0px; --v-progress-linear-height: 2px;\\" role=\\"progressbar\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\">
            <!---->
            <div class=\\"v-progress-linear__background\\" style=\\"width: 100%;\\"></div>
            <transition-stub>
              <div class=\\"v-progress-linear__indeterminate\\">
                <div class=\\"v-progress-linear__indeterminate long\\"></div>
                <div class=\\"v-progress-linear__indeterminate short\\"></div>
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <div class=\\"v-field__field\\" data-no-activator=\\"\\"><label class=\\"v-label v-field-label v-field-label--floating\\" aria-hidden=\\"true\\">
            <!---->TXC-Dateien
          </label><label class=\\"v-label v-field-label\\" for=\\"input-1\\">
            <!---->TXC-Dateien
          </label><input type=\\"file\\" multiple=\\"\\" id=\\"input-1\\" accept=\\".tcx\\" outlined=\\"\\">
          <!---->
        </div>
        <transition-stub>
          <div class=\\"v-field__clearable\\" style=\\"display: none;\\"><i class=\\"mdi-close-circle mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable\\" aria-hidden=\\"true\\"></i></div>
        </transition-stub>
        <div class=\\"v-field__append-inner\\">
          <!---->
          <!---->
        </div>
        <div class=\\"v-field__outline\\">
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <div class=\\"v-input__details\\">
      <transition-group-stub tag=\\"div\\" class=\\"v-messages\\">
        <!---->
      </transition-group-stub>
      <!----><span></span>
      <transition-stub>
        <div class=\\"v-counter\\" style=\\"display: none;\\">0 files</div>
      </transition-stub>
    </div>
  </div>
</div>"
`;
